File: src/application/ports/PayoutProcessor.ts
import { PayoutDestination } from '../../domain/PayoutDestination.js';
import { TransferQuote } from '../../domain/TransferQuote.js';

export interface PayoutProcessor {
  execute(request: {
    walletId: string;
    destination: PayoutDestination;
    quote: TransferQuote;
    reference: string;
  }): Promise<{ payoutId: string; status: 'pending' | 'completed' }>;
}

File: internal/application/ports/payout_processor.go
package ports

import (
	"context"

	"github.com/example/payment-gateway/internal/domain"
)

// PayoutRequest captures the information required to hand off a payout to an external processor.
type PayoutRequest struct {
	WalletID     string
	ExternalID   string
	Destination  domain.PayoutDestination
	SourceAmount domain.Money
	TargetAmount domain.Money
	ExchangeRate string
	Metadata     map[string]string
}

// PayoutProcessor pushes payouts to an infrastructure layer.
type PayoutProcessor interface {
	Process(ctx context.Context, request PayoutRequest) (string, error)
}
